stackslib/src/net/inv/nakamoto.rs

```diff
@@ -105,33 +105,149 @@ impl InvTenureInfo {
+            let parent_id_opt =
+                NakamotoChainState::get_nakamoto_parent_block_id(chainstate.db(), &cursor)?;
+            let Some(parent_id) = parent_id_opt else {
+                return Ok(None);
+            };
+            if self.processed_tenures.contains_key(&parent_id) {
+                return Ok(Some(parent_id));
+            }
+            cursor = parent_id;
         }
+        Ok(None)
     }
 
-    /// Get a processed tenure. If it's not cached, then load it.
-    /// Returns Some(..) if there existed a tenure-change tx for this given consensus hash
-    fn get_processed_tenure(
+    /// Get a processed tenure. If it's not cached, then load it from disk.
+    ///
+    /// Loading it is expensive, so once loaded, store it with the cached processed tenure map
+    /// associated with `tip_block_id`.
+    ///
+    /// If there is no such map, then see if a recent ancestor of `tip_block_id` is represented. If
+    /// so, then remove that map and associate it with `tip_block_id`.  This way, as the blockchain
+    /// advances, cached tenure information for the same Stacks fork stays associated with that
+    /// fork's chain tip (assuming this code gets run sufficiently often relative to the
+    /// advancement of the `tip_block_id` tip value).
+    ///
+    /// Returns Ok(Some(..)) if there existed a tenure-change tx for this given consensus hash
+    /// Returns Ok(None) if not
+    /// Returns Err(..) on DB error
+    pub(crate) fn get_processed_tenure(
         &mut self,
         chainstate: &StacksChainState,
         tip_block_id: &StacksBlockId,
         tenure_id_consensus_hash: &ConsensusHash,
     ) -> Result<Option<InvTenureInfo>, NetError> {
-        // TODO: MARF-aware cache
-        // not cached so go load it
-        let loaded_info_opt =
-            InvTenureInfo::load(chainstate, tip_block_id, &tenure_id_consensus_hash)?;
-        self.processed_tenures
-            .insert(tenure_id_consensus_hash.clone(), loaded_info_opt.clone());
-        Ok(loaded_info_opt)
+        if self.processed_tenures.get(tip_block_id).is_none() {
+            // this tip has no known table.
+            // does it have an ancestor with a table? If so, then move its ancestor's table to this
+            // tip. Otherwise, make a new table.
+            if let Some(ancestor_tip_id) =
+                self.find_ancestor_processed_tenures(chainstate, tip_block_id)?
+            {
+                let ancestor_tenures = self
+                    .processed_tenures
+                    .remove(&ancestor_tip_id)
+                    .unwrap_or_else(|| {
+                        panic!("FATAL: did not have ancestor tip reported by search");
+                    });
+
+                self.processed_tenures
+                    .insert(tip_block_id.clone(), ancestor_tenures);
+            } else {
+                self.processed_tenures
+                    .insert(tip_block_id.clone(), HashMap::new());
```

Comment:
Should `processed_tenures` garbage collect? Otherwise, it seems like it might grow unbounded with the number of forks (which admittedly should be small, but its probably worth doing some limited GC).

Reply:
It can't, because nodes can ask for any inventory vector from any reward cycle (and must do so if they are booting up).

We don't garbage-collect in the epoch 2.x inventory generator either, for the same reason.  It would only lead to thrashing when nodes ask for old inventories.